src/core/env.ml

```diff
@@ -78,15 +78,17 @@ let to_ctxt : env -> ctxt =
   List.map
     (fun (_,(v,a,t)) -> (v, Bindlib.unbox a, Option.map Bindlib.unbox t))
 
-(** [match_prod c t f] returns [f a b] if [t] matches [Prod(a,b)] possibly
-   after reduction. *)
-let match_prod : ctxt -> term -> (term -> tbinder -> 'a) -> 'a = fun c t f ->
+(** [match_prod s c t f] returns [f a b] if [t] matches [Prod(a,b)] possibly
+   after reduction.
+@raise [Invalid_argument s] if [t] is not a product. *)
+let match_prod : string -> ctxt -> term -> (term -> tbinder -> 'a) -> 'a =
```

Comment:
What is the first argument used for?

Reply:
Isn't it only useful for debugging?

Reply:
Right, it helps locate where the invalid argument is introduced.

Reply:
IMO it's better to let people instrument the code when debugging, for instance using `Printexc` to obtain stack traces. I'm afraid adding extra arguments for debugging may be troubling for people discovering the  code.

Reply:
I don't know that. Could you please give more details?

Reply:
Looking at <https://ocaml.org/api/Printexc.html>, you can use `Printexc.record_backtrace true` to (obviously) record backtraces. Consider the following input (with ocaml toplevel)
```ocaml
# let f () = raise Not_found;;
val f : unit -> 'a = <fun>
# let g () = f ();;
val g : unit -> 'a = <fun>
# g ();;
Exception: Not_found.
```
then with backtrace recording:
```ocaml
# Printexc.record_backtrace true;;
- : unit = ()
# g ();;
Exception: Not_found.
Raised at f in file "//toplevel//", line 1, characters 11-26
Called from Toploop.load_lambda in file "toplevel/toploop.ml", line 212, characters 17-27
```

Reply:
That's interesting but note that it works for ocaml >= 3.11.0 only.

Reply:
However this is already used in ls_lsp.ml. So why CI does not fail?

Reply:
Ocaml 3.11 is super old, we're at 4.11 now

Reply:
Right, my mistake : - )

Reply:
Fixed in last commit.